core/src/repair_service.rs

```diff
@@ -601,4 +859,277 @@ mod test {
+            &UdpSocket::bind("0.0.0.0:0").unwrap(),
+        );
+        assert_eq!(duplicate_slot_repair_statuses.len(), 1);
+        assert!(duplicate_slot_repair_statuses.get(&dead_slot).is_some());
+
+        // Insert rest of shreds. Slot is full, should get filtered from
+        // `duplicate_slot_repair_statuses`
+        blockstore
+            .insert_shreds(vec![shreds.pop().unwrap()], None, false)
+            .unwrap();
+        RepairService::generate_and_send_duplicate_repairs(
+            &mut duplicate_slot_repair_statuses,
+            &cluster_slots,
+            &blockstore,
+            &serve_repair,
+            &mut RepairStats::default(),
+            &UdpSocket::bind("0.0.0.0:0").unwrap(),
+        );
+        assert!(duplicate_slot_repair_statuses.is_empty());
+    }
+
+    #[test]
+    pub fn test_update_duplicate_slot_repair_addr() {
+        let dummy_addr = Some(UdpSocket::bind("0.0.0.0:0").unwrap().local_addr().unwrap());
+        let cluster_info = Arc::new(ClusterInfo::new_with_invalid_keypair(
+            Node::new_localhost().info,
+        ));
+        let serve_repair = ServeRepair::new(cluster_info.clone());
+        let valid_repair_peer = Node::new_localhost().info;
+
+        // Signal that this peer has completed the dead slot, and is thus
+        // a valid target for repair
+        let dead_slot = 9;
+        let cluster_slots = ClusterSlots::default();
+        cluster_slots.insert_node_id(dead_slot, Arc::new(valid_repair_peer.id));
+        cluster_info.insert_info(valid_repair_peer);
+
+        // Not enough time has passed, should not update the
+        // address
+        let mut duplicate_status = DuplicateSlotRepairStatus {
+            start: std::u64::MAX,
+            repair_addr: dummy_addr,
+        };
+        RepairService::update_duplicate_slot_repair_addr(
+            dead_slot,
+            &mut duplicate_status,
+            &cluster_slots,
+            &serve_repair,
+        );
+        assert_eq!(duplicate_status.repair_addr, dummy_addr);
+
+        // If the repair address is None, should try to update
+        let mut duplicate_status = DuplicateSlotRepairStatus {
+            start: std::u64::MAX,
+            repair_addr: None,
+        };
+        RepairService::update_duplicate_slot_repair_addr(
+            dead_slot,
+            &mut duplicate_status,
+            &cluster_slots,
+            &serve_repair,
+        );
+        assert!(duplicate_status.repair_addr.is_some());
+
+        // If sufficient time has passssed, should try to update
+        let mut duplicate_status = DuplicateSlotRepairStatus {
+            start: timestamp() - MAX_DUPLICATE_WAIT_MS as u64,
+            repair_addr: dummy_addr,
+        };
+        RepairService::update_duplicate_slot_repair_addr(
+            dead_slot,
+            &mut duplicate_status,
+            &cluster_slots,
+            &serve_repair,
+        );
+        assert!(duplicate_status.repair_addr.is_some());
```

Comment:
maybe `assert_ne!(duplicate_status.repair_addr, dummy_addr);` for more clarity? I bet this assertion would pass even if we don't do timeout handling.

Reply:
ah very true here, good catch!